### ISSUE TITLE
Thread testing fail randomly

### ISSUE THREAD

**Issue by ahartikainen** [stan-dev/pystan#529]
This test fails randomly (1/100?)

```
def test_threading_support(self):
        # Dont test with Windows
        if sys.platform.startswith("win"):
            return
        # Set up environmental variable
        os.environ['STAN_NUM_THREADS'] = "2"
        # Enable threading
        extra_compile_args = ['-pthread', '-DSTAN_THREADS']
        stan_code = """
        functions {
          vector bl_glm(vector mu_sigma, vector beta,
                        real[] x, int[] y) {
            vector[2] mu = mu_sigma[1:2];
            vector[2] sigma = mu_sigma[3:4];
            real lp = normal_lpdf(beta | mu, sigma);
            real ll = bernoulli_logit_lpmf(y | beta[1] + beta[2] * to_vector(x));
            return [lp + ll]';
          }
        }
        data {
          int<lower = 0> K;
          int<lower = 0> N;
          vector[N] x;
          int<lower = 0, upper = 1> y[N];
        }
        transformed data {
          int<lower = 0> J = N / K;
          real x_r[K, J];
          int<lower = 0, upper = 1> x_i[K, J];
          {
            int pos = 1;
            for (k in 1:K) {
              int end = pos + J - 1;
              x_r[k] = to_array_1d(x[pos:end]);
              x_i[k] = y[pos:end];
              pos += J;
            }
          }
        }
        parameters {
          vector[2] beta[K];
          vector[2] mu;
          vector<lower=0>[2] sigma;
        }
        model {
          mu ~ normal(0, 2);
          sigma ~ normal(0, 2);
          target += sum(map_rect(bl_glm, append_row(mu, sigma),
                                 beta, x_r, x_i));
        }
        """
        stan_data = dict(
            K = 4,
            N = 12,
            x = [1.204, -0.573, -1.35, -1.157,
                 -1.29, 0.515, 1.496, 0.918,
                 0.517, 1.092, -0.485, -2.157],
            y = [1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1]
        )
        stan_model = pystan.StanModel(
            model_code=stan_code,
            extra_compile_args=extra_compile_args
        )
        fit = stan_model.sampling(data=stan_data, chains=2, n_jobs=1)
        self.assertIsNotNone(fit)
        fit2 = stan_model.sampling(data=stan_data, chains=2, n_jobs=2)
        self.assertIsNotNone(fit2)
        draw = fit.extract(pars=fit.model_pars+['lp__'], permuted=False)
        lp = {key : values[-1, 0] for key, values in draw.items() if key == 'lp__'}['lp__']
        draw = {key : values[-1, 0] for key, values in draw.items() if key != 'lp__'}
        draw = fit.unconstrain_pars(draw)
        self.assertEqual(fit.log_prob(draw), lp)
        draw2 = fit2.extract(pars=fit2.model_pars+['lp__'], permuted=False)
        lp2 = {key : values[-1, 0] for key, values in draw2.items() if key == 'lp__'}['lp__']
        draw2 = {key : values[-1, 0] for key, values in draw2.items() if key != 'lp__'}
        draw2 = fit2.unconstrain_pars(draw2)
        self.assertEqual(fit2.log_prob(draw2), lp2)
```

Error msg here:
```
FAIL: test_threading_support (pystan.tests.test_extra_compile_args.TestExtraCompileArgs)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda3/envs/linux_gcc_py2/lib/python2.7/site-packages/pystan/tests/test_extra_compile_args.py", line 120, in test_threading_support
    self.assertEqual(fit2.log_prob(draw2), lp2)
AssertionError: -12.356066987384029 != -12.356066987384031
```

**Comment by riddell-stan** [comment]
Interesting. It's the numerical difference that's the problem, right? 
Maybe there's some bug with map_rect? It's strange that it would only 
error 1 out of 100 times.

**Comment by ahartikainen** [comment]
It could even less than that (frequency of failing), but maybe the sum order changes sometimes which could manifest as this kind of behaviour.

**Comment by ahartikainen** [comment]
Actually we run those tests 16 per run --> 1/1000, not sure but have had that a few times.

**Comment by riddell-stan** [comment]
Could be this: https://discourse.mc-stan.org/t/parallel-stl/6469/7

Can we make a test with map_rect which isn't non-deterministic? :)

**Comment by ahartikainen** [comment]
It is probably that.

Is 10 loops enough?

https://github.com/stan-dev/pystan/pull/531

**Comment by ahartikainen** [comment]
Fixed by #531
